### ISSUE TITLE
mindreader should hash one-block files when writing to local filesystem

### ISSUE THREAD

**Issue by matthewdarwin** [dfuse-io/dfuse-eosio#56]
mindreader should hash one-block files when writing to local filesystem. 

Some filesystems don't like it if you write 10k+ files into a single directory, so maybe hash files into subdirectories 5k at a time or so.

of course the merger would need to know how to handle this setup as well.

**Comment by matthewdarwin** [comment]
For testing, we put 300k one-block files in a single folder on zfs fliesystem, without issues.

However for people using ext filesystem, this issue may still be an issue.

**Comment by matthewdarwin** [comment]
This is also needed when writing to minio.  From minio devs:

It is fairly easy for MinIO to introduce a hierarchy internally by hashing the object keys. We chose to not do it for these reasons:
* Applications will never be export to their filesystem or other storage targets because they won't be able to handle a large flat namespace. 
* MinIO's per-directory limit object limit is determined by the number of drives per erasure-set. This is 16 times the limit of the underlying filesystem.
* Keep the backend human-readable for supportability.

**Comment by abourget** [comment]
You mean hash the *whole* thing? or some parts of it?

Their names are used for a few things, to ensure quick contiguousness (with prev_id, and block_id), and most importantly, they are used for sorting.. if the name was a pure hash, it wouldn't be sorted, and that's a basic assumption of the  merger's query with prefix-scans.

**Comment by matthewdarwin** [comment]
I mean like instead of a file like bucketname/0014999900.dbin.zst the file is instead bucketname/00/14/99/0014999900.dbin.zst  Or however you like so then this is less than 5000 files per directory.

When using ext4, you can only get 32K files in minio before it stops adding new ones.

**Comment by abourget** [comment]
Yes, that I agree.  That would apply to merged blocks files, as well as one-block files.

Let me rename this issue.

**Comment by abourget** [comment]
Ok, updated the issue, please comment if that's not what you meant.

**Comment by abourget** [comment]
Those are the places concerned with this: 
* In the `node-manager`, that's for the mindreader-driven bundler (when passed `--mindreader-merge-and-store-directly`) https://github.com/dfuse-io/node-manager/blob/develop/mindreader/merge_archiver.go#L145
* In the merger:
    * https://github.com/dfuse-io/merger/blob/develop/storage.go#L70 (that's in the sort of quick detection algo, to simulate a `scan` with some `prefix` searches)
    * The function defined here: https://github.com/dfuse-io/merger/blob/develop/storage.go#L136
    * Where it's actually used: https://github.com/dfuse-io/merger/blob/develop/merger.go#L390
* In `bstream`, assumed to be such here: https://github.com/dfuse-io/bstream/blob/develop/filesource.go#L134

Ideally, it could be backwards compatible!  That `filesource.go` file could first check with `000/000/0000.dbin.zstd`, and then fallback on `0000000000.dbin.zstd`, perhaps with a ENV var set, otherwise, not even bother.

What would be the best format? `00/00/000000` (? `000/000/0000` ? on EOSIO, it's 10 digits, you can survive the length of the chain).

**Comment by abourget** [comment]
We could have a decorator for the `Store` abstraction, that automatically tweaks the files, on the fly, perhaps when passed with some option like: `s3://blahblah/path?folderize=aa/bb/cccccccccccccccc` or something. This could be declared on the CLI, and added only when needed, and the Store wrapper would handle it both ways.
